### PR TITLE
Make some analytics call conditional

### LINKED ISSUE #4752: Make some analytics call conditional
It'd be useful to have a way to make analytics calls conditional on a value returned by amp:access, so that certain calls are only made for certain users.

Although analytics will wait for the result of an access call before firing if you include any variables from access, it does so even if the value is undefined. Would you ever consider making a given analytics call never fire if a particular amp:access returned undefined for a given variable?

i.e. `requests: { foo: 'http://foo.com?x=AUTHDATA(donuts)' }` would only make call foo if donuts was true or false but not if it was undefined.

### PR DESCRIPTION
**Pull request by zikas**
Added the ability to fire analytics triggers conditional based on `enabled` property that can be specified in individual triggers and at the root level.

FIXES #4752

### CODE REVIEW

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 1]
```diff
@@ -554,23 +554,31 @@ export class AmpAnalytics extends AMP.BaseElement {
...
-        const whiteList = this.isSandbox_ ? SANDBOX_AVAILABLE_VARS : undefined;
-        // For consistency with amp-pixel we also expand any url replacements.
-        return urlReplacementsForDoc(this.element).expandAsync(
-            request, undefined, whiteList);
-      })
-      .then(request => {
-        this.sendRequest_(request, trigger);
-        return request;
-      });
+    return this.checkTriggerEnabled_(trigger, event)
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/amp-analytics.js]: would it make more sense to check for `enabled` before we do the url expansion. That way we prevent some work that might not be used.

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 2]
```diff
@@ -605,6 +612,57 @@ export class AmpAnalytics extends AMP.BaseElement {
   }
 
   /**
+   * Checks if request for a trigger is enabled.
+   * @param {!JSONType} trigger The config to use to determine sampling.
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/amp-analytics.js]: s/sampling/if the trigger is enabled

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 3]
```diff
@@ -605,6 +612,57 @@ export class AmpAnalytics extends AMP.BaseElement {
...
+    return Promise.all([enabledOnTagLevel, enabledOnTriggerLevel])
+        .then(enabled => {
+          dev().assert(enabled.length === 2);
+          return enabled[0] && enabled[1];
+        });
+  }
+
+  /**
+   * Checks result of 'enabled' spec evaluation. Returns false if spec is provided and value
+   * resolves to empty string.
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/amp-analytics.js]: it returns false for all falsey values. right? (eg: 0, null and so on)

File: extensions/amp-analytics/0.1/test/test-amp-analytics.js [hunk 4]
```diff
@@ -1203,6 +1203,148 @@ describe('amp-analytics', function() {
...
+            'on': 'visible',
+            'request': 'pageview1',
+            'vars': {},
+          },
+        },
+        'vars': {},
+      };
+    }
+
+    it('allows a request through', () => {
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/test/test-amp-analytics.js]: allows a request through for undefined \`enabled\` property

File: extensions/amp-analytics/0.1/test/test-amp-analytics.js [hunk 5]
```diff
@@ -1203,6 +1203,148 @@ describe('amp-analytics', function() {
...
+      });
+    });
+
+    it('allows a request based on url-replacements', () => {
+      const config = getConfig();
+      config.triggers.conditional.enabled = '${pageViewId}';
+      const analytics = getAnalyticsTag(config);
+
+      const urlReplacements = urlReplacementsForDoc(analytics.element);
+      sandbox.stub(urlReplacements.getVariableSource(), 'get').returns(0);
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/test/test-amp-analytics.js]: this is interesting. Does the 0 get converted to a string? how does the tigger get enabled for this case?

File: extensions/amp-analytics/0.1/test/test-amp-analytics.js [hunk 6]
```diff
@@ -1203,6 +1203,148 @@ describe('amp-analytics', function() {
...
+
+      const urlReplacements = urlReplacementsForDoc(analytics.element);
+      sandbox.stub(urlReplacements.getVariableSource(), 'get').returns(null);
+
+      return waitForNoSendRequest(analytics).then(() => {
+        expect(sendRequestSpy).to.have.not.been.called;
+      });
+    });
+  });
+
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/test/test-amp-analytics.js]: Can you add a test or two for the case where both tag level and trigger level properties are specified?

File: extensions/amp-analytics/0.1/amp-analytics.js (hunk 3 above)
Comment (zikas) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Method itself always returns Promise<boolean>, so it's real true/false all the time.

In current implementation false is returned only if the result of a spec expansion is an empty string (assuming that url expansion always returns string, the only falsey value is ''). 
Should I change this and treat all falsey-looking strings ('0', 'false', 'undefined', 'null, 'NaN'') as false?

File: extensions/amp-analytics/0.1/test/test-amp-analytics.js (hunk 5 above)
Comment (zikas) [inline extensions/amp-analytics/0.1/test/test-amp-analytics.js]: The result of expandUrlAsync  is always a string because it gets encoded every time, that's the reason 0 (that gets translated to '0') is treated as true

**Comment by zikas** [comment]
ptal

File: extensions/amp-analytics/0.1/test/test-amp-analytics.js (hunk 5 above)
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/test/test-amp-analytics.js]: Knowing the fact that URL replacements does conversion to strings, we still can fairly easily standardize on what strings lead to `false`, including `0` and `false`.

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 7]
```diff
@@ -538,6 +538,42 @@ export class AmpAnalytics extends AMP.BaseElement {
       return Promise.resolve();
     }
 
+    return this.checkTriggerEnabled_(trigger, event)
+        .then(enabled => {
+          if (!enabled) {
+            return;
+          }
+          return this.expandExtraUrlParams_(trigger, event)
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Could you move this into a separate method, e.g. `expandAndSendRequest_`?

File: extensions/amp-analytics/0.1/amp-analytics.js (hunk 3 above)
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/amp-analytics.js]: I think we should go ahead and change this, yes. The only more desirable alternative is to change var/url-replacements expansion with a new mode that will take false-y values and instead of outputting them via `String(x)` would instead output them as an empty string. However, that'd be a lot of changes for a relatively small feature. So, in the meantime, using falsy strings is a good alternative, I believe.

**Comment by dvoytenko** [comment]
@zikas Thanks a lot for the PR! This looks very good. I just added a couple of comments, most importantly around the point you've made on falsy strings.

File: extensions/amp-analytics/0.1/test/test-amp-analytics.js (hunk 6 above)
Comment (zikas) [inline extensions/amp-analytics/0.1/test/test-amp-analytics.js]: Done

File: extensions/amp-analytics/0.1/amp-analytics.js (hunk 7 above)
Comment (zikas) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Done

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 8]
```diff
@@ -538,6 +537,53 @@ export class AmpAnalytics extends AMP.BaseElement {
...
+          }
+          return this.expandAndSendRequest_(request, trigger, event);
+        });
+  }
+
+  /**
+   * @param {string} request The request to process.
+   * @param {!JSONType} trigger JSON config block that resulted in this event.
+   * @param {!Object} event Object with details about the event.
+   * @return {!Promise<string|undefined>} The request that was sent out.
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/amp-analytics.js]: I believe type is always `!Promise<string>` here. no?

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 9]
```diff
@@ -30,9 +30,9 @@ import {isEnumValue} from '../../../src/types';
 import {Activity} from './activity-impl';
 import {Cid} from './cid-impl';
 import {
-    InstrumentationService,
-    instrumentationServicePromiseForDoc,
-    AnalyticsEventType,
+  InstrumentationService,
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/amp-analytics.js]: nit: were the indentation changes automatically generated?

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 10]
```diff
@@ -587,16 +618,15 @@ export class AmpAnalytics extends AMP.BaseElement {
...
     const threshold = parseFloat(spec['threshold']); // Threshold can be NaN.
     if (threshold >= 0 && threshold <= 100) {
-      const keyPromise = this.variableService_.expandTemplate(
-          spec['sampleOn'], this.expansionOptions_({}, trigger))
-        .then(key => urlReplacementsForDoc(this.element).expandAsync(key));
-      return keyPromise
+      const expansionOptions = this.expansionOptions_({}, trigger);
+      return this.expandTemplateWithUrlParams_(sampleOn, expansionOptions)
           .then(key => this.cryptoService_.uniform(key))
           .then(digest => digest * 100 < spec['threshold']);
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Could you please change here `spec['threshold']` to `threshold` from above `const threshold`?

Comment (zikas) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Done

File: extensions/amp-analytics/0.1/amp-analytics.js (hunk 8 above)
Comment (zikas) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Done

File: extensions/amp-analytics/0.1/amp-analytics.js (hunk 9 above)
Comment (zikas) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Yeah, I guess this is my ide. Lint accepts either way. I rolled back unrelated indentation changes.

**Comment by dvoytenko** [comment]
Thanks again! Looks like everything is green. Merging...
